planning/behavior_path_planner/autoware_behavior_path_bidirectional_traffic_module/src/oncoming_car.cpp

```diff
@@ -0,0 +1,276 @@
+  return speed_;
+}
+
+void CarObject::update(const geometry_msgs::msg::Pose & pose, const double & speed)
+{
+  pose_ = pose;
+  speed_ = speed;
+}
+
+std::optional<CarObject> find(const std::vector<CarObject> & oncoming_cars, const size_t & key)
+{
+  auto it = std::find_if(oncoming_cars.begin(), oncoming_cars.end(), [key](const CarObject & car) {
+    return uuid_to_key(car.get_uuid().uuid) == key;
+  });
+
+  if (it != oncoming_cars.end()) {
+    return *it;
+  }
+
+  return std::nullopt;
+}
+
+std::optional<autoware_perception_msgs::msg::PredictedObject> find(
+  const autoware_perception_msgs::msg::PredictedObjects & car_objects, const size_t & key)
+{
+  auto it = std::find_if(
+    car_objects.objects.begin(), car_objects.objects.end(),
+    [key](const auto & object) { return uuid_to_key(object.object_id.uuid) == key; });
+
+  if (it != car_objects.objects.end()) {
+    return *it;
+  }
+
+  return std::nullopt;
+}
+
+template <typename T>
+std::optional<std::pair<size_t, T>> find(const std::map<size_t, T> & map, const size_t & key)
+{
+  auto it = map.find(key);
+  if (it != map.end()) {
+    return *it;
+  }
+
+  return std::nullopt;
+}
+
+void erase(std::vector<CarObject> * oncoming_cars, const size_t & key)
+{
+  oncoming_cars->erase(
+    std::remove_if(
+      oncoming_cars->begin(), oncoming_cars->end(),
+      [&key](const CarObject & oncoming_car) {
+        return uuid_to_key(oncoming_car.get_uuid().uuid) == key;
+      }),
+    oncoming_cars->end());
+}
+
+double distance_on_lane(
+  const geometry_msgs::msg::Pose & pose1, const geometry_msgs::msg::Pose & pose2,
+  const ConnectedBidirectionalLanelets::SharedConstPtr & bidirectional_lanelets)
+{
+  auto center_line = bidirectional_lanelets->get_center_line();
+  double distance = experimental::trajectory::closest(center_line, pose2) -
+                    experimental::trajectory::closest(center_line, pose1);
+  return distance;
+}
+
+bool PredictedObjectsComparator::operator()(
+  const autoware_perception_msgs::msg::PredictedObject & lhs,
+  const autoware_perception_msgs::msg::PredictedObject & rhs) const
+{
+  return std::lexicographical_compare(
+    lhs.object_id.uuid.begin(), lhs.object_id.uuid.end(), rhs.object_id.uuid.begin(),
+    rhs.object_id.uuid.end());
+}
+
+CarObject::CarObject(
+  geometry_msgs::msg::Pose pose, const double & speed, unique_identifier_msgs::msg::UUID uuid,
+  ConnectedBidirectionalLanelets::SharedConstPtr bidirectional_lanelets)
+: pose_(pose),
+  speed_(speed),
+  uuid_(uuid),
+  bidirectional_lanelets_(std::move(bidirectional_lanelets))
+{
+}
+
+void OncomingCars::update(
+  const autoware_perception_msgs::msg::PredictedObjects & predicted_objects,
+  const geometry_msgs::msg::Pose & ego_pose, const rclcpp::Time & time)
+{
+  autoware_perception_msgs::msg::PredictedObjects current_car_objects =
+    extract_car_objects(predicted_objects);
+
+  std::vector<size_t> keys_to_erase;
+
+  for (const auto & object : current_car_objects.objects) {
```

Comment:
As remainder mainly for me:
```suggestion
  // add new candidate/active oncoming vehicle and promote candidate to active
  for (const auto & object : current_car_objects.objects) {
```